File: libs/metor-fsw-2/python/metor-build/metor_build/_wheel.py
"""Minimal wheel writer (stdlib only).

Writes editable wheels (one ``.pth`` payload) and real wheels (arbitrary
files with modes, entry points, dependency metadata, a platform tag) by
hand, so ``build-system.requires`` stays empty and builds work offline.
"""

import base64
import csv
import hashlib
import io
import os
import re
import zipfile

_GENERATOR = "metor-build 0.1"


def _dist(name: str) -> str:
    """The distribution name as it appears in file names (PEP 427/503)."""
    return re.sub(r"[-_.]+", "_", name).lower()


def metadata_contents(
    name: str,
    version: str,
    requires: list[str] | None = None,
    requires_python: str | None = None,
) -> str:
    lines = [f"Metadata-Version: 2.1\nName: {name}\nVersion: {version}\n"]
    if requires_python:
        lines.append(f"Requires-Python: {requires_python}\n")
    for req in requires or []:
        lines.append(f"Requires-Dist: {req}\n")
    return "".join(lines)


def write_metadata(metadata_directory: str, name: str, version: str) -> str:
    """Write ``<dist>-<ver>.dist-info/METADATA``; returns the dist-info name."""
    dist_info = f"{_dist(name)}-{version}.dist-info"
    path = os.path.join(metadata_directory, dist_info)
    os.makedirs(path, exist_ok=True)
    with open(os.path.join(path, "METADATA"), "w", encoding="utf-8") as f:
        f.write(metadata_contents(name, version))
    return dist_info


def write_wheel(
    wheel_directory: str,
    name: str,
    version: str,
    files: list[tuple[str, bytes, int]],
    *,
    tag: str = "py3-none-any",
    requires: list[str] | None = None,
    requires_python: str | None = None,
    entry_points: str | None = None,
) -> str:
    """Write ``<dist>-<ver>-<tag>.whl`` from ``(arcname, data, mode)`` files.

    Modes ride the zip entries' external attributes, which installers
    preserve — how a packaged binary stays executable. Returns the wheel
    file name.
    """
    dist = _dist(name)
    dist_info = f"{dist}-{version}.dist-info"
    records: list[tuple[str, str, str]] = []
    payload: list[tuple[str, bytes, int]] = []

    def entry(path: str, data: bytes, mode: int = 0o644) -> None:
        digest = base64.urlsafe_b64encode(hashlib.sha256(data).digest()).rstrip(b"=")
        records.append((path, f"sha256={digest.decode()}", str(len(data))))
        payload.append((path, data, mode))

    for path, data, mode in files:
        entry(path, data, mode)
    entry(
        f"{dist_info}/METADATA",
        metadata_contents(name, version, requires, requires_python).encode(),
    )
    entry(
        f"{dist_info}/WHEEL",
        (
            "Wheel-Version: 1.0\n"
            f"Generator: {_GENERATOR}\n"
            f"Root-Is-Purelib: {'true' if tag.endswith('-any') else 'false'}\n"
            f"Tag: {tag}\n"
        ).encode(),
    )
    if entry_points is not None:
        entry(f"{dist_info}/entry_points.txt", entry_points.encode())

    record = io.StringIO()
    writer = csv.writer(record, lineterminator="\n")
    writer.writerows(records)
    writer.writerow((f"{dist_info}/RECORD", "", ""))
    payload.append((f"{dist_info}/RECORD", record.getvalue().encode(), 0o644))

    wheel_name = f"{dist}-{version}-{tag}.whl"
    with zipfile.ZipFile(
        os.path.join(wheel_directory, wheel_name), "w", zipfile.ZIP_DEFLATED
    ) as zf:
        for path, data, mode in payload:
            info = zipfile.ZipInfo(path)
            info.external_attr = (0o100000 | mode) << 16
            info.compress_type = zipfile.ZIP_DEFLATED
            zf.writestr(info, data)
    return wheel_name


def write_editable_wheel(wheel_directory: str, name: str, version: str, pth_content: str) -> str:
    """Write ``<dist>-<ver>-py3-none-any.whl`` whose payload is one ``.pth``.

    The ``.pth`` carries absolute path lines, which both the interpreter and
    pyright follow — no import hooks. Returns the wheel file name.
    """
    return write_wheel(
        wheel_directory,
        name,
        version,
        [(f"_{_dist(name)}_editable.pth", pth_content.encode(), 0o644)],
    )

File: libs/metor-fsw-2/src/wiring/wheel.rs
//! A deterministic wheel (zip) writer, the sibling of [`bundle`](super::bundle)'s
//! reproducible tar: entries sorted by archive name, timestamps fixed at the
//! DOS epoch, unix modes in the central directory's external attributes, and
//! every entry **stored** rather than compressed — identical inputs produce
//! byte-identical wheels with no compressor in the loop, and installers
//! accept stored entries per the zip and wheel specs. Pack payloads are a
//! few MB of shared objects; a size guardrail is phase 4's concern
//! (`docs/design-packaging.md` §7.1).
//!
//! [`write_wheel`] assembles the `dist-info` (`METADATA`, `WHEEL`, a sha256
//! `RECORD`) around the caller's files and writes
//! `<dist>-<version>-<tag>.whl`.

use std::io::Write as _;
use std::path::{Path, PathBuf};

use base64::Engine as _;
use sha2::{Digest, Sha256};

/// The wheel's identity and dependency metadata.
#[derive(Clone, Debug)]
pub struct WheelMeta {
    /// The distribution name as authored (`adcs-pack`); file names use the
    /// PEP 427/503 normalization.
    pub dist_name: String,
    /// The distribution version.
    pub version: String,
    /// The wheel tag (`py3-none-any` for a fat pack wheel).
    pub tag: String,
    /// The `Requires-Python` specifier, if any.
    pub requires_python: Option<String>,
    /// `Requires-Dist` lines: the injected pins plus the pack's own
    /// `[project] dependencies`.
    pub requires: Vec<String>,
}

/// One payload file: archive name, contents, and unix mode (`0o644` for
/// text, `0o755` for the shared objects — installers extract the mode from
/// the central directory, which is how a packaged binary stays executable).
#[derive(Clone, Debug)]
pub struct WheelFile {
    /// The archive path, `/`-separated.
    pub arcname: String,
    /// The file contents.
    pub bytes: Vec<u8>,
    /// The unix permission bits.
    pub mode: u32,
}

/// The distribution name as it appears in file names (PEP 427/503): runs of
/// `-`, `_`, `.` collapse to one `_`, lowercased.
pub fn normalize_dist(name: &str) -> String {
    let mut out = String::with_capacity(name.len());
    let mut in_sep = false;
    for c in name.chars() {
        if c == '-' || c == '_' || c == '.' {
            if !in_sep {
                out.push('_');
            }
            in_sep = true;
        } else {
            out.push(c.to_ascii_lowercase());
            in_sep = false;
        }
    }
    out
}

/// Write `<dist>-<version>-<tag>.whl` into `out_dir` and return its path.
pub fn write_wheel(
    out_dir: &Path,
    meta: &WheelMeta,
    files: Vec<WheelFile>,
) -> std::io::Result<PathBuf> {
    let dist = normalize_dist(&meta.dist_name);
    let dist_info = format!("{dist}-{}.dist-info", meta.version);

    let mut entries = files;
    entries.push(WheelFile {
        arcname: format!("{dist_info}/METADATA"),
        bytes: metadata_contents(meta).into_bytes(),
        mode: 0o644,
    });
    entries.push(WheelFile {
        arcname: format!("{dist_info}/WHEEL"),
        bytes: format!(
            "Wheel-Version: 1.0\nGenerator: metor-fsw {}\nRoot-Is-Purelib: {}\nTag: {}\n",
            env!("CARGO_PKG_VERSION"),
            meta.tag.ends_with("-any"),
            meta.tag,
        )
        .into_bytes(),
        mode: 0o644,
    });
    // Stable byte order before RECORD, whose own line order must match.
    entries.sort_by(|a, b| a.arcname.cmp(&b.arcname));

    let mut record = String::new();
    for entry in &entries {
        let digest = Sha256::digest(&entry.bytes);
        let b64 = base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(digest);
        record.push_str(&format!(
            "{},sha256={},{}\n",
            entry.arcname,
            b64,
            entry.bytes.len()
        ));
    }
    record.push_str(&format!("{dist_info}/RECORD,,\n"));
    entries.push(WheelFile {
        arcname: format!("{dist_info}/RECORD"),
        bytes: record.into_bytes(),
        mode: 0o644,
    });

    let name = format!("{dist}-{}-{}.whl", meta.version, meta.tag);
    let path = out_dir.join(name);
    std::fs::write(&path, zip_stored(&entries))?;
    Ok(path)
}

/// The `METADATA` text: core identity plus the dependency lines.
fn metadata_contents(meta: &WheelMeta) -> String {
    let mut out = format!(
        "Metadata-Version: 2.1\nName: {}\nVersion: {}\n",
        meta.dist_name, meta.version
    );
    if let Some(rp) = &meta.requires_python {
        out.push_str(&format!("Requires-Python: {rp}\n"));
    }
    for req in &meta.requires {
        out.push_str(&format!("Requires-Dist: {req}\n"));
    }
    out
}

// ---------------------------------------------------------------------------
// The stored-only zip encoder
// ---------------------------------------------------------------------------

// 1980-01-01 00:00:00, the earliest DOS timestamp — the fixed value that
// makes archives reproducible.
const DOS_DATE: u16 = 0x0021;
const DOS_TIME: u16 = 0;

/// Encode `entries` (already sorted) as a stored-only zip.
fn zip_stored(entries: &[WheelFile]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut central = Vec::new();
    let mut count: u16 = 0;

    for entry in entries {
        let crc = crc32fast::hash(&entry.bytes);
        let name = entry.arcname.as_bytes();
        let size = entry.bytes.len() as u32;
        let offset = out.len() as u32;

        // Local file header.
        out.write_all(&0x04034b50u32.to_le_bytes()).unwrap();
        out.write_all(&20u16.to_le_bytes()).unwrap(); // version needed
        out.write_all(&0u16.to_le_bytes()).unwrap(); // flags
        out.write_all(&0u16.to_le_bytes()).unwrap(); // method: stored
        out.write_all(&DOS_TIME.to_le_bytes()).unwrap();
        out.write_all(&DOS_DATE.to_le_bytes()).unwrap();
        out.write_all(&crc.to_le_bytes()).unwrap();
        out.write_all(&size.to_le_bytes()).unwrap(); // compressed
        out.write_all(&size.to_le_bytes()).unwrap(); // uncompressed
        out.write_all(&(name.len() as u16).to_le_bytes()).unwrap();
        out.write_all(&0u16.to_le_bytes()).unwrap(); // extra len
        out.write_all(name).unwrap();
        out.write_all(&entry.bytes).unwrap();

        // Central directory entry.
        central.write_all(&0x02014b50u32.to_le_bytes()).unwrap();
        // Made by: unix (3) << 8 | zip spec 2.0 — carries the external attrs.
        central.write_all(&0x031eu16.to_le_bytes()).unwrap();
        central.write_all(&20u16.to_le_bytes()).unwrap();
        central.write_all(&0u16.to_le_bytes()).unwrap();
        central.write_all(&0u16.to_le_bytes()).unwrap();
        central.write_all(&DOS_TIME.to_le_bytes()).unwrap();
        central.write_all(&DOS_DATE.to_le_bytes()).unwrap();
        central.write_all(&crc.to_le_bytes()).unwrap();
        central.write_all(&size.to_le_bytes()).unwrap();
        central.write_all(&size.to_le_bytes()).unwrap();
        central
            .write_all(&(name.len() as u16).to_le_bytes())
            .unwrap();
        central.write_all(&0u16.to_le_bytes()).unwrap(); // extra
        central.write_all(&0u16.to_le_bytes()).unwrap(); // comment
        central.write_all(&0u16.to_le_bytes()).unwrap(); // disk
        central.write_all(&0u16.to_le_bytes()).unwrap(); // internal attrs
        let external = (0o100000 | entry.mode) << 16; // regular file + mode
        central.write_all(&external.to_le_bytes()).unwrap();
        central.write_all(&offset.to_le_bytes()).unwrap();
        central.write_all(name).unwrap();
        count += 1;
    }

    let cd_offset = out.len() as u32;
    let cd_size = central.len() as u32;
    out.extend_from_slice(&central);

    // End of central directory.
    out.write_all(&0x06054b50u32.to_le_bytes()).unwrap();
    out.write_all(&0u16.to_le_bytes()).unwrap(); // disk
    out.write_all(&0u16.to_le_bytes()).unwrap(); // cd disk
    out.write_all(&count.to_le_bytes()).unwrap();
    out.write_all(&count.to_le_bytes()).unwrap();
    out.write_all(&cd_size.to_le_bytes()).unwrap();
    out.write_all(&cd_offset.to_le_bytes()).unwrap();
    out.write_all(&0u16.to_le_bytes()).unwrap(); // comment len
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn meta() -> WheelMeta {
        WheelMeta {
            dist_name: "Adcs-Pack".into(),
            version: "1.2.0".into(),
            tag: "py3-none-any".into(),
            requires_python: Some(">=3.11".into()),
            requires: vec!["metor-fsw-abi==8".into(), "metor-config>=0.3,<0.4".into()],
        }
    }

    fn files() -> Vec<WheelFile> {
        vec![
            WheelFile {
                arcname: "adcs_pack/_libs/x/libx.so".into(),
                bytes: vec![1, 2, 3],
                mode: 0o755,
            },
            WheelFile {
                arcname: "adcs_pack/__init__.py".into(),
                bytes: b"# module\n".to_vec(),
                mode: 0o644,
            },
        ]
    }

    #[test]
    fn normalization_is_pep503() {
        assert_eq!(normalize_dist("Adcs-Pack"), "adcs_pack");
        assert_eq!(normalize_dist("a.b--c_d"), "a_b_c_d");
    }

    /// Identical inputs produce byte-identical wheels, regardless of the
    /// order the payload files were supplied in.
    #[test]
    fn wheels_are_reproducible() {
        let dir = tempfile::tempdir().unwrap();
        let (a_dir, b_dir) = (dir.path().join("a"), dir.path().join("b"));
        std::fs::create_dir_all(&a_dir).unwrap();
        std::fs::create_dir_all(&b_dir).unwrap();
        let a = write_wheel(&a_dir, &meta(), files()).unwrap();
        let mut reversed = files();
        reversed.reverse();
        let b = write_wheel(&b_dir, &meta(), reversed).unwrap();
        assert_eq!(
            a.file_name(),
            Some("adcs_pack-1.2.0-py3-none-any.whl".as_ref())
        );
        assert_eq!(std::fs::read(&a).unwrap(), std::fs::read(&b).unwrap());
    }

    /// Python's own zip machinery accepts the archive, sees the fixed
    /// timestamps and modes, and the RECORD digests verify.
    #[test]
    #[cfg(not(miri))]
    fn python_zipfile_reads_it_back() {
        let dir = tempfile::tempdir().unwrap();
        let wheel = write_wheel(dir.path(), &meta(), files()).unwrap();
        let check = std::process::Command::new("python3")
            .arg("-c")
            .arg(
                r#"
import base64, hashlib, sys, zipfile
zf = zipfile.ZipFile(sys.argv[1])
assert zf.testzip() is None
names = zf.namelist()
assert names[:-1] == sorted(names[:-1]), names
assert names[-1].endswith("RECORD"), names
record = zf.read("adcs_pack-1.2.0.dist-info/RECORD").decode()
for line in record.strip().splitlines():
    path, digest, _size = line.rsplit(",", 2)
    if not digest:
        continue
    algo, b64 = digest.split("=", 1)
    actual = base64.urlsafe_b64encode(hashlib.new(algo, zf.read(path)).digest()).rstrip(b"=").decode()
    assert actual == b64, path
info = zf.getinfo("adcs_pack/_libs/x/libx.so")
assert (info.external_attr >> 16) & 0o777 == 0o755
assert info.date_time == (1980, 1, 1, 0, 0, 0)
meta = zf.read("adcs_pack-1.2.0.dist-info/METADATA").decode()
assert "Requires-Dist: metor-fsw-abi==8" in meta
assert "Requires-Python: >=3.11" in meta
print("ok")
"#,
            )
            .arg(&wheel)
            .output()
            .expect("python3 runs");
        assert!(
            check.status.success(),
            "{}",
            String::from_utf8_lossy(&check.stderr)
        );
    }
}
